coil-base/src/main/java/coil/request/Parameters.kt

```diff
@@ -0,0 +1,73 @@
+package coil.request
+
+import androidx.collection.arrayMapOf
+import coil.decode.Decoder
+import coil.fetch.Fetcher
+import coil.util.toArrayMap
+
+/** A map of generic values that can be used to pass custom data to [Fetcher]s and [Decoder]s. */
+class Parameters private constructor(
+    private val map: Map<String, Entry>
+) : Map<String, Parameters.Entry> by map {
+
+    companion object {
+        @JvmField
+        val EMPTY = Parameters()
+
+        /** Create a new [Parameters] instance. */
+        inline operator fun invoke(
+            builder: Builder.() -> Unit = {}
+        ): Parameters = Builder().apply(builder).build()
+    }
+
+    override fun equals(other: Any?) = map == other
+
+    override fun hashCode() = map.hashCode()
+
+    override fun toString() = map.toString()
+
+    fun newBuilder() = Builder(this)
+
+    /**
+     * @param value The parameter's value.
+     * @param cacheKey The parameter's cache key. If not null, this value will be added to a request's cache key.
+     */
+    data class Entry(
+        val value: Any,
```

Comment:
Would it be right to allow value to be nullable too? I guess if we are modeling this off of something like Retrofit and query params, `null` would not be allowed, but I could see some ways in which you might want to be able to pass null to signify that you don't want something to be set, such as if a decoder would automatically perform some behavior and then you want to `null` that out to not perform the automatic behavior. 

In the end, it is a design decision though, so I leave it up to you.

Reply:
Good point. I can't think of a reason to not allow null so probably better to be more permissive. `Decoder`s will have to handle null anyways if the expected parameter is absent.